### ISSUE TITLE
CLA Assistant Lite GitHub Action 

### ISSUE THREAD

**Issue by ibakshay** [cla-assistant/cla-assistant#483]
Hello Community,

We are in development of CLA Assistant Lite Github action (a light version of CLA Assistant). With this Action, we can get rid of the need of a centrally managed database and rather storing all the contributors data in something like  a `cla.md` file.  I understand, this action won't be able to provide all the feature-sets  of the current `CLA Assistant OAuth App` , However, this will work just fine If you want to configure CLA Assistant for a repo and store all the contributors data inside  the repo file system itself. 

## Advantages
1. decentralizing the database of contributors
1. fully integrated with github environment 
1. no UI is required
1. no need to grant any permission/scope

So, The CLA Lite Action will be triggered whenever there is a  `pull_request` event and will check If the Contributors have  signed the CLA. 

## Signing CLA inside GitHub Environment
Now the question arises on how the contributors will  sign the CLA **inside the GitHub environment**. What we thought is to ask the contributors to [react]( https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) with 👍 on the CLA Pull request comment for signing the CLA something like the below screenshot.

![image](https://user-images.githubusercontent.com/33329946/64777327-37b6a380-d55a-11e9-97f8-e186b1181bed.png)

Please kindly let us know your thoughts and comments on CLA Lite GitHub Action or If you have any better solution for signing the CLA inside the GitHub environment . 

cc: @thojansen  @KharitonOff

**Comment by daveaglick** [comment]
Love this idea!

> What we thought is to ask the contributors to react with 👍 on the CLA Pull request comment

While that seems like a great user experience and I'd prefer it without any other considerations, I'm a little worried about how easy it is and whether it adequately expresses an intent to "sign" the CLA.

What about requiring a comment with specific wording like "I, Dave Glick, agree to the terms of this CLA". You could provide a template in the comment from the bot:

> Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our Contributor License Agreement before we can accept your contribution.
>
> Please add a comment to this issue with the below text to acknowledge and sign the CLA:
>
> I, [Your Name], agree to the terms of the provided CLA

I also noticed the bot comment in the screen shot above doesn't actually link to the CLA - I'm assuming that was just omitted while testing out the language?

**Comment by germanattanasio** [comment]
I was looking into deploying this in a database I own because the original app is sometimes slow, but I agree with all your points, especially removing the need for the UI or database.

**Comment by thojansen** [comment]
@daveaglick Agree that a comment would be more specific. This could even enable custom fields, but would require a clear syntax, e.g. put a comment which ends with something like the following.

Hereby, I sign v1 of the  CLA with Name: Thomas Jansen, E-Mail: my@email.com

**Comment by paddyfeerick** [comment]
One idea I had to record signing within the GH UI is to do so by raising a Pull request on a repo holding the CLA text itself. (You edit the text in a specific, precise way to signal your agreement - for example,  add 'YES' after 'I agree'). The PR then triggers processing to:
- confirm the text was otherwise unaltered
- (GH has gpg-signed the commit so you know who it is!)
- make a further edit (and commit) to restore the text to its original form
- merges the PR. (Or abandoned the PR if the text was not correct)

Thus you have a record of who has signed in the commit history of a simple repo.

**Comment by thojansen** [comment]
We had some iterations where the CLA itself should be stored (in the same repo, separate repo, ..) and we still think a versioned gist file is a good tradeoff. Also not sure if we want to enforce signed commits and how commonly this feature is used right now. I think this should definitely be optional.

So the CLA is already versioned with timestamps and the signing always refers to a specific version.

**Comment by paddyfeerick** [comment]
Of course the commits made through the GH UI are signed, as that's just Github doing its thing. Nothing to stop someone committing outside GH of course. In which case there is no obligation to sign.

**Comment by ibakshay** [comment]
Hello @germanattanasio , @daveaglick  , @germanattanasio  
Unfortunately, there is only read access for the GitHub Actions when there is a Pull Request coming from the forked repository for security reasons.  Hence, The CLA Action cannot even post a pull request comment when there is a pull request from the forked repo. I have raised this issue in many GitHub Community Discussions like [here](https://github.community/t5/GitHub-Actions/Token-permissions-for-forks-once-again/m-p/34160#M1819) , [here](https://github.community/t5/GitHub-Actions/Make-secrets-available-to-builds-of-forks/m-p/33885/highlight/true#M1696) and [here](https://github.community/t5/GitHub-Actions/Github-Workflow-not-running-from-pull-request-from-forked/m-p/33547/highlight/true#M1555) and I haven't got any positive response from the GitHub side. 
However, Many people put forward the proposal for having read/write access for the forked repository, Since, this will directly affect **ALL** the code auditing actions like  [coveralls](https://github.com/coverallsapp/github-action), [labeller](https://github.com/actions/labeler/issues/12), unit tests and I am optimistic that GitHub will come up with some solution or workaround to counter this issue..  Like allowing Read/write access to the PR coming from the forked repository for **Trusted Actions** ? . 

The only way to proceed forward now with these restrictions is to develop CLA Action in such a way that, The CLA Action will run as a [CRON](https://help.github.com/en/articles/events-that-trigger-workflows#scheduled-events-schedule) job for every 15 Mins or so (depending upon  the rate Limit) to do the CLA check.. The GitHub Actions is free for the Open Source Projects, Hence running CLA Actions as a CRON job won't be a problem.  Please kindly share your thoughts and feedback for this approach 😄 .

**Comment by daveaglick** [comment]
The security restriction makes sense for arbitrary action, but hopefully they come up with some kind of trusted flag or other way to mark the action as having access.

In the meantime, I'd be fine with allowing the action to poll. PRs almost always require some iteration and presumably the submitter will be watching their notifications or revisiting the PR and will see the CLA messages even if they're not added immediately.

**Comment by ibakshay** [comment]
Hi @daveaglick @paddyfeerick  @germanattanasio, We have developed `CLA Assistant github action` and it is in beta stage. Feel free to test and give us your feedback 😄 . 
You can find the `cla assistant github action` [here](https://github.com/cla-assistant/github-action)

**Comment by silverhook** [comment]
I have given the whole “how to automate and simplify signing of CLAs” bit quite a bit of thought before, and came to the conclusion that relying on Git + SSH keys (perhaps with a `--signed-off-by` flag as with the DCO) should be perfectly fine and the lowest level you can automate it.

- in almost all jurisdictions even for an exclusive CLA the formalities require a written form, and for a full copyright assignment a signature (see [Comparative Analysis of Copyright Assignment and License Formalities for Open Source Contributor Agreements](http://script-ed.org/wp-content/uploads/2013/08/rens.pdf) by Andres Guadamuz and Andrew Rens
- in almost all jurisdictions a digital signature is equivalent to a pen signature (ibidem)
- an SSH key (pair) should qualify as a digital signature – I’m pretty sure it does in the EU, I would be surprised if it did not elsewhere – especially since in the context of Git(Hub) it is associated with an e-mail address

So if I make a commit that includes a statement that for this commit/contribution I am signing the relevant CLA (perhaps most easily wrapped in a similar system as the DCO), and use my SSH key to push the commit to a repo on GitHub, the record in git history with my name, e-mail address and the fact that I agreed with the CLA and used my SSH key to do so, should be enough to meet the legal requirements of both a written agreement (CLA text in repo and referenced) and the signature (SSH key associated with my name and e-mail address).

**Comment by gnarea** [comment]
I think it'd be good to have the CLA and the signatories stored in a well-known location per organisation (not per repo). For example, GitHub already uses `.github` repos, so we could default to things like `<org>/.github/cla-assistant/cla.json` and `<org>/.github/cla-assistant/signatories.json`.

@silverhook, I haven't used SSH keys on GitHub for ages, since I use IntelliJ products + personal access tokens, so your proposed solution wouldn't work me unless I create an SSH key just to do the signature.

**Comment by silverhook** [comment]
@gnarea, true, the SSH key solution has that limitation.

On one hand it’s legally closer to an actual signature, on the other hand, it does carry that additional burden. Which is smaller than the burden of pen and paper, but still present.

But using the git `-s` flag even without SSH might be good enough as well (depending on the content of the CLA/CAA and jurisdiction). In any case it is easier to link with the committer signing off than relying only on a content of a file.

One problem I see with the `.github/` location is that it’s GitHub-specific. IMHO it would be better to make this process forge-agnostic and rely simply on Git.

**Comment by Olf0** [comment]
Oh my, it started off so nicely by [agreeing to a CLA simply by hitting "☺→:+1:"](https://github.com/cla-assistant/cla-assistant/issues/483#issue-491727700) (although the link to the CLA proper was missing in the example text), but went to over-engineering in the second post and ultimately ended in suggesting to enforce the use of SSH keys and cryptographic signatures (which are only linguistically related to written signatures, but not legally).

Thus unfortunately the ["contributor assistant" GH-action](https://github.com/contributor-assistant/github-action) has become much more complicated to use (luckily still without cryptographic signatures) than the technically dreaded [CLA assistant](https://cla-assistant.io/), which requires just one or two mouse-clicks.

Mind that one has to be logged in at GitHub to react to a comment by ":+1:", hence authenticity is not an issue.  The ":+1:" also is a clear indication of will, especially as this is unambiguously denoted as the requested reaction in the comment text.  The only issue I can see, is that a ":+1:"-reaction can usually be reverted at any time: While revoking a CLA should be possible, both transactions, the signing and the revoking of a CLA must be immutably recorded.
<br />

#### TL;DR
The [original scheme](https://github.com/cla-assistant/cla-assistant/issues/483#issue-491727700) was perfect in terms of usability and legally on par with the extant [CLA assistant](https://cla-assistant.io/), but the ["contributor assistant" GH-action](https://github.com/contributor-assistant/github-action) has become an over-engineered monster with low usability.

Hence please return to the simplicity of the [original scheme](https://github.com/cla-assistant/cla-assistant/issues/483#issue-491727700) (and of the [CLA assistant](https://cla-assistant.io/)), requiring one or two clicks but nothing more.
